paymentsheet/src/test/java/com/stripe/android/paymentelement/confirmation/DefaultConfirmationHandlerTest.kt

```diff
@@ -1,2169 +1,891 @@
-                    ),
-                    config = GOOGLE_PAY_OPTION.config.copy(
-                        merchantCurrencyCode = null,
-                    )
-                ),
+    private fun ConfirmationHandler.Result.assertCanceled(): ConfirmationHandler.Result.Canceled {
+        assertThat(this).isInstanceOf<ConfirmationHandler.Result.Canceled>()
+
+        return this as ConfirmationHandler.Result.Canceled
+    }
+
+    private class Scenario(
+        val confirmationHandler: DefaultConfirmationHandler,
+        val errorReporter: FakeErrorReporter,
+        val savedStateHandle: SavedStateHandle,
+        val someDefinitionScenario: RecordingConfirmationDefinition.Scenario<
+            SomeConfirmationDefinition.Option,
+            SomeConfirmationDefinition.Launcher,
+            SomeConfirmationDefinition.LauncherArgs,
+            SomeConfirmationDefinition.LauncherResult,
+            >,
+        val someDefinitionOnResult: ((SomeConfirmationDefinition.LauncherResult) -> Unit)?,
+        val someOtherDefinitionScenario: RecordingConfirmationDefinition.Scenario<
+            SomeOtherConfirmationDefinition.Option,
+            SomeOtherConfirmationDefinition.Launcher,
+            SomeOtherConfirmationDefinition.LauncherArgs,
+            SomeOtherConfirmationDefinition.LauncherResult,
+            >,
+        val someOtherDefinitionOnResult: ((SomeOtherConfirmationDefinition.LauncherResult) -> Unit)?,
+    )
+
+    private class SomeConfirmationDefinition(
+        action: ConfirmationDefinition.Action<LauncherArgs> =
+            ConfirmationDefinition.Action.Fail(
+                cause = IllegalStateException("Failed!"),
+                message = R.string.stripe_something_went_wrong.resolvableString,
+                errorType = ConfirmationHandler.Result.Failed.ErrorType.Internal,
             ),
-        )
-
-        val result = defaultConfirmationHandler.awaitResult().asFailed()
-
-        val message = "GooglePayConfig.currencyCode is required in order to use " +
-            "Google Pay when processing a Setup Intent"
-
-        assertThat(result.cause.message).isEqualTo(message)
-        assertThat(result.type).isEqualTo(ConfirmationHandler.Result.Failed.ErrorType.MerchantIntegration)
-        assertThat(result.message).isEqualTo(R.string.stripe_something_went_wrong.resolvableString)
-
-        assertThat(logger.getLoggedMessages()).contains(message)
-    }
+        result: ConfirmationDefinition.Result = ConfirmationDefinition.Result.Canceled(
+            action = ConfirmationHandler.Result.Canceled.Action.InformCancellation,
+        ),
+    ) : FakeConfirmationDefinition<
+        SomeConfirmationDefinition.Option,
+        SomeConfirmationDefinition.Launcher,
+        SomeConfirmationDefinition.LauncherArgs,
+        SomeConfirmationDefinition.LauncherResult,
+        >(
+        launcher = Launcher,
+        action = action,
+        result = result,
+    ) {
```

Comment:
Naming is hard lol. Considered also naming these `FirstConfirmationDefinition` & `SecondConfirmationDefinition` but that felt like they are meant to be ordered as such. Open to other suggestions.